capsules/src/segger_rtt.rs

```diff
@@ -0,0 +1,199 @@
+pub static mut DOWN_BUFFER: [u8; 32] = [0; 32];
+
+/// This structure is defined by the segger RTT protocol. It must exist in
+/// memory in exactly this form so that the segger JTAG tool can find it in the
+/// chip's memory and read and write messages to the appropriate buffers.
+#[repr(C)]
+pub struct SeggerRttMemory {
+    id: [u8; 16],
+    number_up_buffers: u32,
+    number_down_buffers: u32,
+    up_buffer: SeggerRttBuffer,
+    down_buffer: SeggerRttBuffer,
+}
+
+#[repr(C)]
+pub struct SeggerRttBuffer {
+    name: *const u8, // Pointer to the name of this channel. Must be a 4 byte thin pointer.
+    buffer: *const u8, // Pointer to the buffer for this channel.
+    length: u32,
+    write_position: u32,
+    read_position: u32,
+    flags: u32,
+}
+
+impl SeggerRttMemory {
+    pub fn new(
+        up_buffer_name: &'static [u8],
+        up_buffer: &'static [u8],
+        down_buffer_name: &'static [u8],
+        down_buffer: &'static [u8],
+    ) -> SeggerRttMemory {
+        SeggerRttMemory {
+            // Must be "SEGGER RTT".
+            id: *b"SEGGER RTT\0\0\0\0\0\0",
+            number_up_buffers: 1,
+            number_down_buffers: 1,
+            up_buffer: SeggerRttBuffer {
+                name: up_buffer_name.as_ptr(),
+                buffer: up_buffer.as_ptr(),
+                length: 1024,
+                write_position: 0,
+                read_position: 0,
+                flags: 0,
+            },
+            down_buffer: SeggerRttBuffer {
+                name: down_buffer_name.as_ptr(),
+                buffer: down_buffer.as_ptr(),
+                length: 32,
+                write_position: 0,
+                read_position: 0,
+                flags: 0,
+            },
+        }
+    }
+}
+
+pub struct SeggerRtt<'a, A: hil::time::Alarm + 'a> {
+    alarm: &'a A, // Dummy alarm so we can get a callback.
+    config: TakeCell<'static, SeggerRttMemory>,
+    up_buffer: TakeCell<'static, [u8]>,
+    _down_buffer: TakeCell<'static, [u8]>,
+    client: OptionalCell<&'static hil::uart::Client>,
+    client_buffer: TakeCell<'static, [u8]>,
+}
+
+impl<'a, A: hil::time::Alarm + 'a> SeggerRtt<'a, A> {
+    pub fn new(
+        alarm: &'a A,
+        config: &'static mut SeggerRttMemory,
+        up_buffer: &'static mut [u8],
+        down_buffer: &'static mut [u8],
+    ) -> SeggerRtt<'a, A> {
+        SeggerRtt {
+            alarm: alarm,
+            config: TakeCell::new(config),
+            up_buffer: TakeCell::new(up_buffer),
+            _down_buffer: TakeCell::new(down_buffer),
+            client: OptionalCell::empty(),
+            client_buffer: TakeCell::empty(),
+        }
+    }
+}
+
+impl<'a, A: hil::time::Alarm + 'a> hil::uart::UART for SeggerRtt<'a, A> {
+    fn set_client(&self, client: &'static hil::uart::Client) {
+        self.client.set(client);
+    }
+
+    fn init(&self, _params: hil::uart::UARTParams) {}
+
+    fn transmit(&self, tx_data: &'static mut [u8], tx_len: usize) {
```

Comment:
I don`t really understand how `read and write indexes work` here:

Is this supposed to be a circular buffer? But, Isn`t read always zero?

Also, what happens `length` wraps and shouldn`t we check that `tx_len` <= `length`?